src/t_hash.c

```diff
@@ -1028,10 +1056,401 @@ static void addHashIteratorCursorToReply(writePreparedClient *wpc, hashTypeItera
+        new_argv = zmalloc(sizeof(robj *) * (num_fields + 5));
+        if (set_expired)
+            new_argv[new_argc++] = shared.hdel;
+        else if (persist)
+            new_argv[new_argc++] = shared.hpersist;
+        else
+            new_argv[new_argc++] = shared.hpexpireat;
+
+        new_argv[new_argc++] = c->argv[1];
+        incrRefCount(c->argv[1]);
+
+        if (set_expiry) {
+            new_argv[new_argc++] = NULL; // placeholder for the expiration time
+            milliseconds_index = new_argc - 1;
+        }
+
+        if (set_expiry || persist) {
+            new_argv[new_argc++] = shared.fields;
+            new_argv[new_argc++] = NULL; // placeholder for the number of objects
+            numitems_index = new_argc - 1;
+        }
+    }
+    for (i = fields_index; i < c->argc; i++) {
+        int changed = 0;
+        addHashFieldToReply(c, o, c->argv[i]->ptr);
+        if (set_expired) {
+            changed = hashTypeDelete(o, c->argv[i]->ptr);
+        } else if (set_expiry) {
+            changed = (hashTypeSetExpire(o, c->argv[i]->ptr, when, 0) == 1) ? 1 : 0;
+        } else if (persist) {
+            changed = (hashTypePersist(o, c->argv[i]->ptr) == 1) ? 1 : 0;
+        }
+        if (changed) {
+            changes++;
+            new_argv[new_argc++] = c->argv[i];
+            incrRefCount(c->argv[i]);
+        }
+    }
+
+    /* rewrite the command vector and persist in case there are changes.
+     * Also notify keyspace notifications and signal the key was changed. */
+    if (changes) {
+        if (milliseconds_index > 0) {
+            milliseconds_obj = createStringObjectFromLongLong(when);
+            new_argv[milliseconds_index] = milliseconds_obj;
+            incrRefCount(milliseconds_obj);
+        }
+        if (numitems_index > 0) {
+            numitems_obj = createStringObjectFromLongLong(changes);
+            new_argv[numitems_index] = numitems_obj;
+            incrRefCount(numitems_obj);
+        }
+        replaceClientCommandVector(c, new_argc, new_argv);
+        if (set_expired)
+            notifyKeyspaceEvent(NOTIFY_HASH, "hexpired", c->argv[1], c->db->id);
+        else
+            notifyKeyspaceEvent(NOTIFY_HASH, set_expiry ? "hexpire" : "hpersist", c->argv[1], c->db->id);
+        if (milliseconds_obj) decrRefCount(milliseconds_obj);
+        if (numitems_obj) decrRefCount(numitems_obj);
+
+        server.dirty += changes;
+        signalModifiedKey(c, c->db, c->argv[1]);
+
+        /* Delete the object in case it was left empty */
+        if (hashTypeLength(o) == 0) {
+            dbDelete(c->db, c->argv[1]);
+            notifyKeyspaceEvent(NOTIFY_GENERIC, "del", c->argv[1], c->db->id);
+        }
+    } else {
+        /* If no changes were done we still need to free the new argv array and the refcount of the first argument. */
+        if (set_expiry || set_expired || persist) {
+            decrRefCount(c->argv[1]);
+        }
+        if (new_argv) zfree(new_argv);
```

Comment:
I think this can be simplified to:
```suggestion
        if (set_expiry || set_expired || persist) {
            decrRefCount(c->argv[1]);
            zfree(new_argv);
        }
```
because the `new_argv` allocation only happens when `set_expiry || set_expired || persist` is true.

Reply:
I am too scared to do it (such a complicated function :) ) but I think you are correct

Reply:
actually I thought you were pointing a different location. in this case the new argv is only allocated if any of the optional expiration arguments was provided, which can NOT be the case if you are just using this command to get a value.